OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseTabsDashboardFragment.java

```diff
@@ -139,6 +172,104 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
         }
     }
 
+    private void setupIAPLayout() {
+        binding.accessError.setVisibility(View.VISIBLE);
+        binding.accessError.setState(EdxCourseAccessErrorState.State.IS_UPGRADEABLE);
+        binding.toolbar.tabs.setVisibility(View.GONE);
+        setupToolbar(true);
+
+        iapViewModel = new ViewModelProvider(this).get(InAppPurchasesViewModel.class);
+        boolean isPurchaseEnabled = environment.getAppFeaturesPrefs().isIAPEnabled(environment.getLoginPrefs().isOddUserId());
+        if (isPurchaseEnabled) {
+            initIAPObservers();
+            // Shimmer container taking sometime to get ready and perform the animation, so
+            // by adding the some delay fixed that issue for lower-end devices, and for the
+            // proper animation.
+            binding.accessError.postDelayed(
+                    () -> iapViewModel.initializeProductPrice(courseData.getCourseSku()), 1500);
+        }
+        binding.accessError.setSecondaryButtonListener(onFindCourseClick());
+    }
+
+    private void initIAPObservers() {
+        iapViewModel.getProductPrice().observe(this, skuDetailsEvent -> {
+            SkuDetails skuDetails = skuDetailsEvent.getContentIfNotConsumed();
+            if (skuDetails != null) {
+                setUpUpgradeButton(skuDetails);
+            }
+        });
+
+        iapViewModel.getLaunchPurchaseFlow().observe(this, booleanEvent ->
+                iapViewModel.purchaseItem(requireActivity(), environment.getLoginPrefs().getUserId(),
+                        courseData.getCourseSku())
+        );
+
+        iapViewModel.getShowLoader().observe(this, showLoaderEvent -> {
+            Boolean showLoader = showLoaderEvent.getContentIfNotConsumed();
+            if (showLoader != null) {
+                binding.accessError.enableUpgradeButton(showLoader);
+            }
+        });
+
+        iapViewModel.getErrorMessage().observe(this, errorMessageEvent -> {
+            ErrorMessage errorMessage = errorMessageEvent.getContentIfNotConsumed();
+            if (errorMessage != null) {
+                handleIAPException(errorMessage);
+            }
+        });
+
+        iapViewModel.getProductPurchased().observe(this, iapFlowDataEvent -> {
+            IAPFlowData iapFlowData = iapFlowDataEvent.getContentIfNotConsumed();
+            if (iapFlowData != null) {
+                showFullscreenLoader(iapFlowData);
+            }
+        });
+    }
+
+    private void setUpUpgradeButton(SkuDetails skuDetails) {
+        // The app get the sku details instantly, so add some wait to perform
+        // animation at least one cycle.
+        binding.accessError.setPrimaryButtonText(ResourceUtil.getFormattedString(
+                getResources(),
+                R.string.label_upgrade_course_button,
+                AppConstants.PRICE,
+                skuDetails.getPrice()
+        ).toString());
+
+        binding.accessError.setPrimaryButtonListener(view -> {
```

Comment:
onClick is not working correctly and responding slowly.
also the loader is not visible on button after tapping it.

Reply:
fixed.